crypto/evp/digest.c

```diff
@@ -237,6 +238,15 @@ static int evp_md_init_internal(EVP_MD_CTX *ctx, const EVP_MD *type,
     cleanup_old_md_data(ctx, 1);
 
     /* Start of non-legacy code below */
+    if (ctx->algctx != NULL && ctx->digest != type) {
+        if (!ossl_assert(ctx->digest != NULL)) {
+            ERR_raise(ERR_LIB_EVP, EVP_R_INITIALIZATION_ERROR);
+            return 0;
+        }
+        if (ctx->digest->freectx != NULL)
+            ctx->digest->freectx(ctx->algctx);
+        ctx->algctx = NULL;
+    }
 
     if (type->prov == NULL) {
```

Comment:
If we are doing an implicit fetch (what this block goes on to do), would that risk having the provider change from under us and thereby invalidate any existing algctx?  Or are we guaranteed that if there is an algctx there must already be an associated provider?
In particular, this block can change the value of `type`!  So the check at L262 is testing different things than the check at R241.

Reply:
If there is an algctx set, the ctx->digest is already the fetched (non-implicit) digest (see https://github.com/openssl/openssl/pull/18105/files#diff-df03d3f4aedaf8c344451553a2dededfb98fc9c871899a55e4e07cdb5d7a228eR280). So if the type is an implicit digest it cannot be equal to ctx->digest in the if above and the algctx will be freed.

The removed block below was actually completely bogus and the code was not broken just because we always freed the algctx anyway.

Reply:
Okay, so you are saying that the only way `ctx->algctx` can be set is through the "R280" line linked above, which only occurs if there is already a fetched digest from a provider.  That's a reasonable argument, but I think we also have to consider the `EVP_MD_CTX_copy_ex()` case, though, which is another path to set `ctx->algctx` and seems to allow for further complications.  For example, if we try to make a copy where `out` (already on input to the function) has an algctx set but `in` is legacy, it seems we might be able to create an `EVP_MD_CTX` with non-NULL `algctx` but where `ctx->digest` is a legacy (non-provider) digest, which makes the code in this PR a bit harder to reason about.  (Down in https://github.com/openssl/openssl/blob/6009997abd2594d5a7c0606176f404190922b74d/crypto/evp/digest.c#L566-L568 we only NULL out pointer "in case of error" in the non-legacy case, but logic akin to what you are adding here in this PR in R228 is not present for `EVP_MD_CTX_copy{,_ex}`.)

I would be more comfortable (and able to readily approve a PR) if we were `ossl_assert()`ing that `type->prov` was non-NULL (if `ctx->algctx != NULL`) up around R240 where we have a similar sanity check for `ctx->digest`.  I could try to put together a PR to clean up the legacy case of `EVP_MD_CTX_copy_ex()` if you don't want to try to roll those changes into this PR.

Reply:
I do not see how in the legacy path in EVP_MD_CTX_copy_ex you could end up with non-null algctx even in case of error.

Reply:
I was envisioning something like
```
EVP_MD_CTX *questionable, *provided_ctx, *legacy_ctx;
EVP_MD *provided, *legacy;
ENGINE *e = ...; /* something not NULL */
provided = EVP_MD_fetch(...);
legacy = EVP_sha256();
provided_ctx = EVP_MD_CTX_new();
legacy_ctx = EVP_MD_CTX_new();
EVP_DigestInit_ex(provided_ctx, provided, NULL);
EVP_DigestInit_ex(legacy_ctx, legacy, e);
EVP_MD_CTX_copy_ex(questionable, provided_ctx);
EVP_MD_CTX_copy_ex(questionable, legacy_ctx);
```
As I understand it, the penultimate line will set `questionable->algctx` and the last line will not clear it.
(I have not tested this.)

Reply:
It will be cleared by https://github.com/openssl/openssl/blob/6009997abd2594d5a7c0606176f404190922b74d/crypto/evp/digest.c#L612

Reply:
Okay, I see it now.  Thanks for the pointer.
I would still be more comfortable if we asserted that the state of the object is what we expected, but am willing to approve the change without it.  I will see if I can review the update prompted by Dmitry's review.
(Sorry for the long response times here, I'm travelling at the moment.)